### PR TITLE
*: update proto response, errors, retry

### PR DESCRIPTION
**Pull request by disksing**

### CODE REVIEW

File: error.go [hunk 1]
```diff
@@ -19,24 +19,22 @@ import (
...
-	errGetMissing = errors.New("get request is missing")
 	// errInvalidResponse represents response message is invalid.
 	errInvalidResponse = errors.New("invalid response")
 	// errBodyMissing response body is missing error
 	errBodyMissing = errors.New("response body is missing")
-	// errInvalidDSN is returned when store dsn is invalid.
-	errInvalidDSN = errors.New("invalid tikv dsn")
 )
 
+// TiDB decides whether retry transaction by checking if error message contains
```
Comment (siddontang) [inline error.go]: retires

File: error.go [hunk 2]
```diff
@@ -19,24 +19,22 @@ import (
...
+// string "try again later" literally.
```
Comment (siddontang) [inline error.go]: is only checking string safe?

File: error.go [hunk 3]
```diff
@@ -19,24 +19,22 @@ import (
...
+// In TiClient we use `errors.Annotate(err, txnRetryableMark)` to directive TiDB
```
Comment (siddontang) [inline error.go]: s/directive/direct/

File: error.go [hunk 4]
```diff
@@ -19,24 +19,22 @@ import (
...
+// to restart a transaction.
+// Note that it should be only used if i) the error occurs inside a transaction
+// and ii) the error is not totally unexpected and hopefully will recover soon.
+const txnRetryableMark = "[try again later]"
+
 func errLockInfoMissing(k []byte) error {
 	return errors.Errorf("key is locked but lockInfo not found key[%q]", k)
```
Comment (siddontang) [inline error.go]: `is not found, key `

File: kv.go [hunk 5]
```diff
@@ -160,7 +159,9 @@ func (s *tikvStore) CurrentVersion() (kv.Version, error) {
 	return kv.NewVersion(startTS), nil
 }
 
-// sendKVReq send req to region, if errInnerRetryable occur(like NotLeader) then retry by self.
+// sendKVReq send req to tikv server. It will retry internally to find the right
```
Comment (siddontang) [inline kv.go]: SendKVReq sends

File: error.go (hunk 2 above)
Comment (disksing) [inline error.go]: @siddontang  It prevents TiDB from depending on plugin stores or vice versa. I think it's not very elegant but practical.
It is TiDB's choice, anyway.

File: snapshot_test.go [hunk 6]
```diff
@@ -140,9 +140,9 @@ func (s *testSnapshotSuite) TestBatchGetLock(c *C) {
 		snapshot := newTiKVSnapshot(s.store, kv.Version{Ver: txn3.tid})
 		_, err = snapshot.BatchGet(keys)
 		// rowNum/2 key is locked, failed
-		c.Assert(err, NotNil)
+		c.Assert(err, IsNil)
 
-		time.Sleep(time.Duration(cleanupTimeout) * time.Millisecond)
+		time.Sleep(time.Duration(TTL) * time.Millisecond)
```
Comment (morefreeze) [inline snapshot_test.go]: Is this `Sleep` redundancy? If L141 return err == nil, locks must been cleaned.

File: ticlient_test.go [hunk 7]
```diff
@@ -128,22 +129,15 @@ func (s *testTiclientSuite) TestCleanLock(c *C) {
 	err := txn.Commit()
 	c.Assert(err, IsNil)
 
+	time.Sleep(time.Duration(TTL) * time.Millisecond)
```
Comment (morefreeze) [inline ticlient_test.go]: This `Sleep` is not suitable here. 
If `Commit` without `backoff` it's still successful. But I think you want to prove `backoff` worked.

File: snapshot.go [hunk 8]
```diff
@@ -88,36 +88,32 @@ func (s *tikvSnapshot) doBatchGet(singleBatchGet *batchGetRegion) (map[string][]
...
-	if resType := cmdBatchGetResp.GetResType(); resType == nil || resType.GetType() != pb.ResultType_Ok {
-		log.Errorf("Unexpected result type [%s]", resType)
-		return nil, errors.Trace(errRetryable)
-	}
-	results := cmdBatchGetResp.GetResults()
-	log.Debugf("Receive batchGet results length[%d]", len(results))
-	m := make(map[string][]byte, len(results))
+	pairs := cmdBatchGetResp.GetPairs()
+	m := make(map[string][]byte, len(pairs))
 	// go func() for concurrent.
```
Comment (ngaut) [inline snapshot.go]: Remove stale comments.

**Comment by ngaut** [comment]
PTAL @qiuyesuifeng

File: error.go (hunk 1 above)
Comment (disksing) [inline error.go]: @siddontang Seems it should be 'decides whether to retry ...'

File: error.go (hunk 3 above)
Comment (disksing) [inline error.go]: @siddontang I did want to use 'directive', which means 'order'.

Comment (siddontang) [inline error.go]: directive is not verb.

Comment (disksing) [inline error.go]: aha, thanks :)

**Comment by disksing** [comment]
PTAL @ngaut @morefreeze @qiuyesuifeng @siddontang

File: snapshot.go [hunk 9]
```diff
@@ -162,46 +156,50 @@ func (s *tikvSnapshot) BatchGet(keys []kv.Key) (map[string][]byte, error) {
...
+		if err != nil {
+			return nil, errors.Trace(err)
+		}
+		resp, err := s.store.SendKVReq(req, region)
+		if err != nil {
+			return nil, errors.Trace(err)
+		}
+		if regionErr := resp.GetRegionError(); regionErr != nil {
+			backoffErr = regionBackoff()
+			continue
```
Comment (morefreeze) [inline snapshot.go]: It seems always return StaleEpoch error, so no need `continue` instead returning retryable error directly.

**Comment by morefreeze** [comment]
LGTM

**Comment by ngaut** [comment]
PTAL @siddontang

File: lock.go [hunk 10]
```diff
@@ -39,83 +36,96 @@ func newLock(store *tikvStore, pLock []byte, lockVer uint64, key []byte, ver uin
...
+	)
+	return NewBackoff(maxRetry, sleepBase, sleepCap, EqualJitter)
+}
+
 // TTL time to live
-const TTL uint64 = 300                 // ms
-const cleanupTimeout uint64 = TTL * 10 // ms
-// cleanup cleanup lock.pk
-// if lock.key != nil, call rollback&get or commit&get
+const TTL uint64 = 3000 // ms
```
Comment (shenli) [inline lock.go]: Make this private? Add add more comment to indicate this is TTL for lock.

**Comment by shenli** [comment]
LGTM

**Pull request merged**